File: src/main/frontend/demo/component/popover/popover-interactive-tooltip.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/button';
import '@vaadin/integer-field';
import '@vaadin/popover';
import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import type { PopoverTrigger } from '@vaadin/popover';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('popover-interactive-tooltip')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  @state()
  trigger: PopoverTrigger[] = ['hover', 'focus'];

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-integer-field id="cvv-field" label="CVV" style="width: 60px"></vaadin-integer-field>
      <vaadin-popover
        for="cvv-field"
        .trigger="${this.trigger}"
        position="top"
        theme="arrow"
        aria-labelledby="cvv-heading"
      >
        <h3 id="cvv-heading" style="margin: 0; font-size: 1rem">Card Verification Value</h3>
        <div style="max-width: 300px">
          A three or four digit code, usually printed on the back of the card, next to, or at the
          end of, the signature strip.
        </div>
        <a href="https://www.cvvnumber.com/cvv.html" target="_blank">
          See where to find CVV on different cards
        </a>
      </vaadin-popover>
      <!-- end::snippet[] -->
    `;
  }
}

File: src/main/java/com/vaadin/demo/component/popover/PopoverInteractiveTooltip.java
package com.vaadin.demo.component.popover;

import com.vaadin.flow.component.html.Anchor;
import com.vaadin.flow.component.html.AnchorTarget;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.H3;
import com.vaadin.flow.component.popover.Popover;
import com.vaadin.flow.component.popover.PopoverPosition;
import com.vaadin.flow.component.popover.PopoverVariant;
import com.vaadin.flow.component.textfield.IntegerField;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("popover-interactive-tooltip")
public class PopoverInteractiveTooltip extends Div {

    public PopoverInteractiveTooltip() {
        // tag::snippet[]
        IntegerField cvv = new IntegerField("CVV");
        cvv.setWidth("60px");

        Popover popover = new Popover();
        popover.addThemeVariants(PopoverVariant.ARROW);
        popover.setPosition(PopoverPosition.TOP);
        popover.setOpenOnClick(false);
        popover.setOpenOnHover(true);
        popover.setOpenOnFocus(true);

        H3 header = new H3("Card Verification Value");
        header.setId("cvv-heading");
        header.getStyle().set("margin", "0");
        header.getStyle().set("font-size", "1rem");

        Div content = new Div(
                "A three or four digit code, usually printed on the back of the card, "
                        + "next to, or at the end of, the signature strip.");
        content.getStyle().set("max-width", "300px");

        Anchor link = new Anchor("https://www.cvvnumber.com/cvv.html",
                "See where to find CVV on different cards", AnchorTarget.BLANK);

        popover.add(header, content, link);
        popover.setAriaLabelledBy("cvv-heading");
        popover.setTarget(cvv);

        add(cvv, popover);
        // end::snippet[]
    }

    public static class Exporter extends // hidden-source-line
            DemoExporter<PopoverInteractiveTooltip> { // hidden-source-line
    } // hidden-source-line
}
